Libraries/Opc.Ua.Security.Certificates/X509Certificate/X509PfxUtils.cs

```diff
@@ -0,0 +1,190 @@
+        /// </summary>
+        public static bool VerifyRSAKeyPair(
+            X509Certificate2 certWithPublicKey,
+            X509Certificate2 certWithPrivateKey,
+            bool throwOnError = false)
+        {
+            bool result = false;
+            RSA rsaPrivateKey = null;
+            RSA rsaPublicKey = null;
+            try
+            {
+                // verify the public and private key match
+                rsaPrivateKey = certWithPrivateKey.GetRSAPrivateKey();
+#if NETSTANDARD2_1
+                // on .NET Core 3 
+                rsaPrivateKey.ExportParameters(true);
+#endif
+                rsaPublicKey = certWithPublicKey.GetRSAPublicKey();
+                X509KeyUsageFlags keyUsage = GetKeyUsage(certWithPublicKey);
+                if ((keyUsage & X509KeyUsageFlags.DataEncipherment) != 0)
+                {
+                    result = VerifyRSAKeyPairCrypt(rsaPublicKey, rsaPrivateKey);
+                }
+                else if ((keyUsage & X509KeyUsageFlags.DigitalSignature) != 0)
+                {
+                    result = VerifyRSAKeyPairSign(rsaPublicKey, rsaPrivateKey);
+                }
+                else
+                {
+                    throw new CryptographicException("Don't know how to verify the public/private key pair.");
+                }
+            }
+            catch (Exception)
+            {
+                if (throwOnError)
+                {
+                    throwOnError = false;
+                    throw;
+                }
+            }
+            finally
+            {
+                RsaUtils.RSADispose(rsaPrivateKey);
+                RsaUtils.RSADispose(rsaPublicKey);
+                if (!result && throwOnError)
+                {
+                    throw new CryptographicException("The public/private key pair in the certficates do not match.");
+                }
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Creates a certificate from a PKCS #12 store with a private key.
+        /// </summary>
+        /// <param name="rawData">The raw PKCS #12 store data.</param>
+        /// <param name="password">The password to use to access the store.</param>
+        /// <returns>The certificate with a private key.</returns>
+        public static X509Certificate2 CreateCertificateFromPKCS12(
+            byte[] rawData,
+            string password
+            )
+        {
+            Exception ex = null;
+            int flagsRetryCounter = 0;
+            X509Certificate2 certificate = null;
+
+            // We need to try MachineKeySet first as UserKeySet in combination with PersistKeySet hangs ASP.Net WebApps on Azure
+            X509KeyStorageFlags[] storageFlags = {
+                X509KeyStorageFlags.Exportable | X509KeyStorageFlags.PersistKeySet | X509KeyStorageFlags.MachineKeySet,
+                X509KeyStorageFlags.Exportable | X509KeyStorageFlags.PersistKeySet | X509KeyStorageFlags.UserKeySet
```

Comment:
I hope you don't call this very often...

Reply:
Yeah its a little tricky to get to the private keys on various platforms and this has a history from .NET Core 1.x, UWP Asp.Net etc. Luckily only once to get to the application key...